qt/Window.cpp

```diff
@@ -0,0 +1,159 @@
+/* -*- Mode: C++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4; fill-column: 100 -*- */
+/*
+ * Copyright the Collabora Online contributors.
+ *
+ * SPDX-License-Identifier: MPL-2.0
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include "WebView.hpp"
+#include "bridge.hpp"
+#include "DetachableTabs.hpp"
+
+#include <QApplication>
+#include <QMessageBox>
+#include <QTabWidget>
+#include <QTabBar>
+#include <QWidget>
+#include <QCloseEvent>
+#include <LibreOfficeKit/LibreOfficeKit.h>
+#include "qt.hpp"
+#include "Window.hpp"
+
+Window::Window(QWidget * parent, WebView * owner): QMainWindow(parent), owner_(owner) {}
+
+void Window::setCloseCallback(const std::function<void()>& closeCallback)
+{
+    closeCallback_ = closeCallback;
+}
+
+// Used when + button is clicked to open a new tab
+void Window::plusClicked() {
+    // Try to invoke the web UI for the focused tab, if available.
+    QTabWidget* tabWidget = qobject_cast<QTabWidget*>(this->centralWidget());
+    QTabBar* tabBar = tabWidget ? tabWidget->tabBar() : nullptr;
+    int currentIndex = tabBar ? tabBar->currentIndex() : -1;
+
+    WebView* owner = nullptr;
+    if (currentIndex != -1 && tabWidget) {
+        QWidget* w = tabWidget->widget(currentIndex);
+        if (w) {
+            quint64 ownerPtr = w->property("webview_owner").toULongLong();
+            owner = reinterpret_cast<WebView*>((quintptr)ownerPtr);
+        }
+    }
+
+    if (owner) {
+        // Let the page show the new-document UI
+        owner->runJS("app.map.backstageView.show();");
+        return;
+    }
+
+    // No owner available (placeholder tab or transient state). Create a
+    // new document programmatically using the application profile so we
+    // never rely on the native dialog fallback.
+    WebView* newWebView = WebView::createNewDocument(Application::getProfile(), "writer");
+    if (newWebView)
+        newWebView->activateWindow();
+}
+
+void Window::closeEvent(QCloseEvent * ev) {
+    if (closeCallback_) {
+        closeCallback_();
+    }
+
```

Comment:
One of the commits introduced trailing whitespace here, which a later commit removed again; but which was not ideal when I tried to rebase this commit series onto me local head, as it the first commit caused merge failures, fixing which then had `online`'s git commit hooks kick in and refuse to accept the merge-fixed commit due to the trailing whitespace.  Better force-push the removal of that trailing whitespace into the commit that introduced it.